### ISSUE TITLE
Box can dissapear after publish.

### ISSUE THREAD

**Issue by istarkov** [webstudio-is/webstudio-builder#604]
Add a new Box, change the background color to make it visible, then publish. On the published site it's invisible because of zero height.

This is because in design mode we set  `padding-top`, `padding-right` styles on the designer canvas.

This also prevents set `width` and  `height` less than the padding size (50px).

**Comment by kof** [comment]
What Webflow does:
1. adds wf-empty class that sets padding-right and padding bottom only in the designer to give the div some size
2. if user sets width,height,min-width,min-height, padding, display: grid or inserts a child of any size, even 0, then it removes the padding

**Comment by kof** [comment]
So far we have done the same but we didn't remove the padding when user sets those above named properties

**Comment by kof** [comment]
What I don't like about Webflow's solution is that their default padding is not reflected in the style panel in any way. It's magical. Because of that you can't knowingly remove that padding and you preview/published version is unexpectedly collapsed.

**Comment by kof** [comment]
One way to solve it: actually set padding using preset, so that user can see that value in the style panel and we also explain in the UI where the value comes from. User can either actively change it, reset it or ignore it.

**Comment by taylornowotny** [comment]
Having used Webflow divs heavily, I agree that it is a little confusing and frustrating that they have hidden padding on the div. But what's confusing about it is the fact that the padding is hidden. There's a way to see the hidden padding but I forget how. It's not intuitive.

It is not really an issue that the padding is there - it was more of an issue that the padding is hidden so a new user doesn't know where the size of the div is coming from. After I figured it out, I understood that divs need to have some minimum size so you can select them on the canvas and drag elements inside them.

It would be annoying if I had to deliberately change the padding to get those preset styles removed. I like that they disappear automatically, I just wish they communicated that padding was there in the first place.

So I suggest we do it just like Webflow but with the padding values visible in the "preset" color style. The component already has these styles ready to go. When a user adds any children, any padding, or any size values the padding should disappear.

**Comment by kof** [comment]
@taylornowotny the decision we need to make if whether padding is supposed to stay on a preview or published site or not.
If it stays - it can be part of preset styles.

We can remove that preset padding (when certain properties are changed by the user) even when it comes from preset since we know if the style comes from user styles or from preset.

But we need to decide if we want to have that padding consistently  across canvas, preview, published site and not just on canvas.

**Comment by kof** [comment]
We are already doing a bit better than webflow by at least showing the padding value in the spacing UI, even though we still have the same problem as webflow - when you preview or publish it's gone.

**Comment by taylornowotny** [comment]
I think we should show it in the preview and published site. Nobody means to have a blank div on their site. So let's avoid the weirdness and just show it. If a user didn't mean to add that div to their site then the preview will help them find it. I don't see a downside to showing it.

**Comment by kof** [comment]
Alright and then we would remove the preset padding when user changes one of these: width,height,min-width,min-height, padding, display: grid or inserts a child of any size,

**Comment by kof** [comment]
Damn, this will lead to this problem:
1. user inserts a child - we remove the preset padding
2. user removes the child - box collapses since there is now no padding

**Comment by kof** [comment]
To make it always work we have to make sure we never end-up with 0-size elements. Also using a shortlist of properties mentioned above to remove the preset padding is a bad logic. There are too many things that can lead to a box having a proper size, e.g. a layout.

We need to only remove the preset padding when the box has actually a 0-size, based on actual client rect.

**Comment by kof** [comment]
Here is a specific bug in Webflow that is the result of such "naive" preset padding implementation based on a shortlist of properties:

I put a flex on a div, set a height of 50px, then inserted a child div, set the flex-grow: 1 on it.

Expected: child is as big as the parent
Result: child is bigger than the parent on canvas because it has 75px padding, because webflow doesn't know flex grow gives div a size and it doesn't need the preset any more.

<img width="1080" alt="Screenshot 2022-12-16 at 13 46 33" src="https://user-images.githubusercontent.com/52824/208112224-b31b08b1-4d3f-4c91-beb4-7de44811d068.png">

**Comment by kof** [comment]
The problem is that the child instance itself doesn't need to change in this case to receive a height, only parent needs to change and I bet this can happen as a result of a change to any of the parents. We would have to react to size changes in any ancestor in the tree.

**Comment by kof** [comment]
I researched container queries. @container requires a container-type property to be set on the container we want to query and as it turned out container-type kills height: auto. It doesn't work when the height is determined by the child elements and breaks the natural flow of the size computation. It works when size is defined by layout or width/height specifically.

**Comment by kof** [comment]
I have just tested the getBoundingClientRect performance by generating 10000 divs with random heights and calling getBoundingClientReact on each of them. On my machine it took 30ms.

Given the fact that we only need to do it when something changes in the ancestor tree, we will only have 1-20 (estimated) calls at all, where 20 is a 20 levels deep nesting, that should be very rare

@rpominov wdyt?

**Comment by rpominov** [comment]
From performance point of view, this may work.

But I'm not sure it's good from UX point of view to constantly tinker with size like this. What if user wants it to collapse? Maybe they want to make a separator line using a collapsed div.

Maybe we should have a toggle in styles panel that would allow to switch off this behaviour for an instance.

I probably even prefer it to switch off automatically:

1. I add a new Box (toggle is on)
2. I set width and height (toggle switches off automatically)
3. I reset width and height (toggle remains off unless I turn it off manually)

**Comment by kof** [comment]
> Maybe they want to make a separator line using a collapsed div.
Yes we would have to take into consideration if some specific properties like height/width were set to 0 on the element itself

**Comment by kof** [comment]
> I set width and height (toggle switches off automatically)

Problem is that its not just width/height, its lots of properties, including those on the parent that would make a div have a height without actually changing the div itself

**Comment by rpominov** [comment]
> Problem is that its not just width/height, its lots of properties, including those on the parent that would make a div have a height without actually changing the div itself

I know. I just used width/height as example. Under the hood we would measure with `getBoundingClientRect`

**Comment by kof** [comment]
I would assume intentional collapsing is a very rare case that could be allowed specifically by setting width/height on the element. This would be straightforward forward and that's how you would do it in code anyways. So yes its a property check, but its specifically 2 properties and nothing else.

**Comment by kof** [comment]
> But then we're back at analysing properties rather than measuring with getBoundingClientRect

Yes partly, but we remove the bugs like the one in the screenshot. Essentially we would be detecting when element was intentionally set to 0 using a shortlist of padding, width/height and min-, but the non-zero size would be detected from any source

**Comment by taylornowotny** [comment]
My thoughts:

It's ok if a div collapses to size 0 sometimes. That's the expected behavior - the preset padding is just a quality-of-life feature.

In Webflow, in the scenario where I put a child in a div, then take it out and the div goes to size 0, this has never been a problem for me. I just select the div from the Navigator.

**Comment by kof** [comment]
@taylornowotny div has no height of its own, it will always be collapsed initially, you def. don't want that

**Comment by taylornowotny** [comment]
> @taylornowotny div has no height of its own, it will always be collapsed initially, you def. don't want that

I'm responding to your earlier message where you said:

> Damn, this will lead to this problem:
> 
>     user inserts a child - we remove the preset padding
>     user removes the child - box collapses since there is now no padding

I don't see this as a problem that we need to prevent. If the div collapses in this scenario the user can select the div from the Navigator as I have done many time in Webflow. No issue

**Comment by kof** [comment]
> I don't see this as a problem that we need to prevent. If the div collapses in this scenario the user can select the div from the Navigator as I have done many time in Webflow. No issue

You have a professional deformation caused by bad ux :) I am certain this is not something any human wants, not even me as a developer

**Comment by kof** [comment]
Things shouldn't magically disappear and appear somewhere else :)

**Comment by taylornowotny** [comment]
I don't see it as disappearing. In this scenario the div is simply acting like a div. The extra padding on new divs is nice, but if it goes away after putting a child inside then taking it out, that's just fine to me.

I'm not saying that's the perfectly ideal UX, just that it's not a big deal. So if there's a big technical challenge to making it work differently, we would be fine if we just did it like Webflow. If you can improve it and you want to then cool, but I don't think it's critical.

**Comment by kof** [comment]
webflow restores padding helper when child gets deleted, you must have had some other edge case

**Comment by rpominov** [comment]
I think we're essentially trying to predict what the user would want and do it automatically. In an environment with many degrees of freedom, where user can do a lot and can want a lot, this can’t be done well.

I'd keep it very simple (similar to what @kof suggested in https://github.com/webstudio-is/webstudio-designer/issues/604#issuecomment-1346667716):

 - When user adds a Box, it has `min-width` and `max-width` set. 
 - Regular “set” or “preset” styles.
 - Not appearing or disappearing under any conditions.
 - Applied on canvas and in the published site as all other styles.
 
And what @taylornowotny says also sounds good to me.

**Comment by kof** [comment]
@rpominov the only problem with min-size is that it only applies to block elements, but maybe we can live with that and have for inline elements some other solution

**Comment by kof** [comment]
Its clear now how webflow ended up relying on a shortlist of properties. Determining if an element is actually collapsed is just so damn inefficient.

**Comment by kof** [comment]
Maybe there isn't a huge problem if we have no automation around removing spacers? E.g. if its min-width 50px and min-height 50px, the question is how often do you actually have block elements that are smaller than that.

Only in case they are smaller you would have to unset the min-value.

**Comment by taylornowotny** [comment]
I am returning to this issue to add my opinion that I think we should have preset padding on new divs and that padding should disappear when any size properties are added and when any children are added.

Adding min width or height is not what I would want or expect as a user.

Design:
<img width="433" alt="Screenshot 2023-02-14 at 9 48 27 AM" src="https://user-images.githubusercontent.com/86495896/218817216-981a9c79-40fa-4265-a093-4d547baaf432.png">

**Comment by kof** [comment]
I still do think that a new user will potentially add backgrounds to the box without changing it's size and then wonder why its gone on a published site

**Comment by taylornowotny** [comment]
I see that it's more complex than I imagined.

So will we be able to keep a UX like Webflow's or do we need to figure something else out?

**Comment by taylornowotny** [comment]
If we can't figure out a way to make the helper collapse that isn't problematic, then maybe the next best thing is to just let divs be divs and let them collapse. New elements are selected automatically anyways, so if a user adds styles immediately then the helper isn't necessary. Beyond that, they can use the navigator to select an empty div.

**Comment by kof** [comment]
I found a way to universally handle this. They will collapse on production if there was nothing that holds them back from collapsing, but we will correctly handle spacing helpers and remove them universally without having a shortlist of properties that remove them and without writing spacers into instance styles permanently.

**Comment by taylornowotny** [comment]
I don't quite understand this solution technically but I'm on board if the UX is good.

What does the UX look like with this fix - like this?
- New boxes have a helper
- Helper is removed when any property is changed 
- Helper doesn't show on preview or publish

**Comment by kof** [comment]
- only boxes that would collapse otherwise have helper
- helper is removed when any CSS property is changed on that instance that would prevent collapsing, so that helper can be removed
- Helper doesn't show on the preview or publish
- Helper goes away if an instance inserted as a child
- There is no need to set padding-right or padding-bottom if you just need a small div with a defined or layout-based size, as soon as div is not collapsing, helper is not applied

**Comment by kof** [comment]
More: padding will be only added on the side that would collapse otherwise. 
For example when I add a div, it is a block element, it grows automatically full width but has 0 height, in this case spacing heloper with padidng-top: 50px will apply, so that it doesn't collapse.
If user sets height: 100px or does anything that would give it a height - we remove the helper padding right away, so user can actually see the height they set

In other words we prevent elements from collapsing when they have 0 height or width by making them non-zero on canvas, but then we remove those padding universally without shortlists as soon as element doesn't collapse.

**Comment by kof** [comment]
I think this is the most natural behavior we can get while keeping element accessible on the canvas.

**Comment by taylornowotny** [comment]
That sounds perfect!
